Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,309 @@
+  have pisurj : Function.Surjective π := by apply surjective_quotient_mk' -- π is surjective
+  set pipi := Prod.map π π
+  have pipiopen : IsOpenMap pipi := IsOpenMap.prod hpiopen hpiopen -- π × π open
+  have pipisurj : (Function.Surjective (pipi) ) :=  -- π × π surj
+    Function.Surjective.Prod_map pisurj pisurj
+  have pipipquotient := -- π × π is q QuotientMap because open, continuous and surj
+    IsOpenMap.to_quotientMap pipiopen (Continuous.prod_map picont picont) pipisurj
+  rw [<-QuotientMap.isClosed_preimage pipipquotient] -- closed iff preimage closed
+  set gr' := pipi ⁻¹' diagonal (Quotient R) -- preimage of the diag
+  set m : G × X → X × X := fun (g,x) => (g • x, x)
+  set gr := range m -- graph of the orbit relation
+  have r_eq_r' : gr' = gr := by -- the preimage of the diag is the graph of the relation
+    ext ⟨x,y⟩
+    conv_lhs => -- we work only on the left hand side for now
+      rw [mem_preimage]
+      rw [mem_diagonal_iff]
+      change (π x = π y)  --↔ (x, y) ∈ gr
+      rw [Quotient.eq']
+      change ((MulAction.orbitRel G X).Rel x y)-- ↔ (x, y) ∈ gr
+      rw [MulAction.orbitRel_apply]
+      rw [MulAction.orbit]
+      change (∃ g : G, g • y = x)
+    conv_rhs =>
+      rw [mem_range]
+      simp
+    simp [m]
+  rw [r_eq_r']
+  exact hproper.isProperMap_smul_pair'.isClosedMap.closed_range
+
+/-- If a T2 group acts properly on a topological space, then this topological space is T2. -/
+@[to_additive "If a T2 group acts properly on a topological space,
+then this topological space is T2."]
+theorem t2Space_of_properSMul_of_t2Group [h_proper : ProperSMul G X] [T2Space G] : T2Space X := by
+  let f := fun x : X ↦ ((1 : G), x)
+  have proper_f : IsProperMap f := by
+    apply isProperMap_of_closedEmbedding
+    rw [closedEmbedding_iff]
+    constructor
+    · let g := fun gx : G × X ↦ gx.2
+      have : Function.LeftInverse g f := fun x ↦ by simp
+      exact this.embedding (by fun_prop) (by fun_prop)
+    · have : range f = ({1} ×ˢ univ) := by simp
+      rw [this]
+      exact isClosed_singleton.prod isClosed_univ
+  rw [t2_iff_isClosed_diagonal]
+  let g := fun gx : G × X ↦ (gx.1 • gx.2, gx.2)
+  have proper_g : IsProperMap g := (properSMul_iff G X).1 h_proper
+  have : g ∘ f = fun x ↦ (x, x) := by ext x <;> simp
+  have range_gf : range (g ∘ f) = diagonal X := by simp [this]
+  rw [← range_gf]
+  exact (proper_f.comp proper_g).closed_range
+
+/-- If two groups `H` and `G` act on a topological space `X` such that `G` acts properly and
+there exists a group homomorphims `H → G` which is a closed embedding compatible with the actions,
+then `H` also acts properly on `X`. -/
+@[to_additive "If two groups `H` and `G` act on a topological space `X` such that `G` acts properly
+and there exists a group homomorphims `H → G` which is a closed embedding compatible with the
+actions, then `H` also acts properly on `X`."]
+theorem properSMul_of_closed_embedding {H : Type*} [Group H] [MulAction H X] [TopologicalSpace H]
```

Comment:
Again, `UpperCamelCase` becomes `lowerCamelCase` in lemma names, not `snake_case`.
```suggestion
theorem properSMul_of_closedEmbedding {H : Type*} [Group H] [MulAction H X] [TopologicalSpace H]
```